robot_modeling/DQ_SerialManipulator.m

```diff
@@ -289,48 +194,18 @@ function set_effector(obj,effector)
             % This function does not take into account any base or
             % end-effector displacements.
             
-            if nargin == 4
-                n = ith;
-                x_effector = obj.raw_fkm(q,ith);
-                J = obj.raw_pose_jacobian(q,ith);
-                vec_x_effector_dot = J*q_dot(1:ith);
+            if nargin == 4 && ith < obj.n_links
+                % If the Jacobian is not related to the mapping between the
+                % end-effector velocities and the joint velocities, it takes
+                % into account only the base displacement
+                J_dot = hamiplus8(obj.reference_frame)*obj.raw_pose_jacobian_derivative(...
+                    q, q_dot, ith);
             else
-                n = obj.n_links;
-                x_effector = obj.raw_fkm(q);
-                J = obj.raw_pose_jacobian(q);
-                vec_x_effector_dot = J*q_dot;
-            end
-                                 
-            x = DQ(1);            
-            J_dot = zeros(8,n);
-
-            for i = 0:n-1
-                % Use the standard DH convention
-                if strcmp(obj.convention,'standard')
-                    w = DQ.k;
-                    z = DQ(obj.get_z(x.q));
-                else % Use the modified DH convention
-                    w = DQ([0,0,-sin(obj.alpha(i+1)),cos(obj.alpha(i+1)),0, ...
-                        0,-obj.a(i+1)*cos(obj.alpha(i+1)),...
-                        -obj.a(i+1)*sin(obj.alpha(i+1))] );
-                    z = 0.5*x*w*x';
-                end
-                
-                % When i = 0 and length(theta) = 1, theta(1,i) returns
-                % a 1 x 0 vector, differently from the expected
-                % behavior, which is to return a 0 x 1 matrix.
-                % Therefore, we have to deal with the case i = 0
-                % explictly.
-                if i ~= 0
-                    vec_zdot = 0.5*(haminus8(w*x') + ...
-                        hamiplus8(x*w)*DQ.C8) * ...
-                        obj.raw_pose_jacobian(q,i)*q_dot(1:i);
-                else
-                    vec_zdot = zeros(8,1);
-                end
-                J_dot(:,i+1) = haminus8(x_effector)*vec_zdot +...
-                    hamiplus8(z)*vec_x_effector_dot;
-                x = x*obj.dh2dq(q(i+1),i+1);
+                % Otherwise, it the Jacobian is related to the
+                % end-effector velocity, it takes into account both base
+                % and end-effector (constant) displacements.
+                J_dot = hamiplus8(obj.reference_frame)*haminus8(obj.effector)*...
```

Comment:
Here my suspicion was confirmed. If `ith == obj.n_links`, the end-effector transformation is **not** supposed to be considered. The reason behind that decision, which was already implemented in previous versions, is that if one wants to consider the end-effector, it suffices to use the method without the `ith` parameter. That provides a mechanism for obtaining the final transformation while considering the base but not the end-effector transformation.

In summary: 
- `raw_method_name`-> neither base transformation nor end-effector transformation
- `method_name` -> both base transformation and end-effector transformation
- `method_name(ith = number_of_links`) -> only base transformation

Reply:
Hi @bvadorno, (assuming you are referring to the internal implementation of `pose_jacobian_derivative()`)

I modifed the line 
```cpp
if nargin == 4 && ith < obj.n_links
```
to 
```cpp
if nargin == 4
```
in `pose_jacobian_derivative()`. Now, when `ith==obj.n_links` the end-effector transformation is not considered. 

Please let me know if this modification is the only one required to accomplish the expected behavior.  

Best regards, 

Juancho

Reply:
Hi @juanjqo,

It is, thanks. Here we have an important lesson: one should never change the behavior of a method without prior permission.

Best wishes,
Bruno

Reply:
Hi @bvadorno, I agree. 

The current implementation of `DQ_SerialManipulator.pose_jacobian_derivative()` in the master branch is implemented as a raw_method(), this is, both the base and the effector are not taken into account. So, I moved that method (In the current PR) to the subclass `DQ_SerialManipulatorDH` and renamed it as `raw_pose_jacobian_derivative()`. Then, I just implemented the `DQ_SerialManipulator.pose_jacobian_derivative()` inspired in the current implementation of `DQ_SerialManipulator.pose_jacobian(),` which includes the second part in the `if` condition.  `(if nargin == 3 && ith < obj.n_links)`.

Current implementation of pose Jacobian (master branch):
```matlab
function J = pose_jacobian(obj, q, ith)
    if nargin == 3 && ith < obj.n_links
        J = hamiplus8(obj.reference_frame)*obj.raw_pose_jacobian(...
            q, ith);
    else
        J = hamiplus8(obj.reference_frame)*haminus8(obj.effector)*...
            obj.raw_pose_jacobian(q);
    end
end
```

To summarize:

- The method DQ_SerialManipulator.pose_jacobian() is implemented (in the master branch and, therefore, in the current PR since I did not modified that method) in a way that if `ith == obj.n_links` the end-effector transformation is considered. 

- The method DQ_SerialManipulator.pose_jacobian_derivative() (in the current PR with the requested modifications) has a different behavior. When `ith == obj.n_links` the end-effector transformation is **not** considered. 

So, I was wondering, if the DQ_SerialManipulator.pose_jacobian() is not working as is supposed to work (for a long time), may I open a ticket to report the bug?

Reply:
Hi @juanjqo,

Your implementation of the `pose_jacobian_derivative` is okay. Regarding the `pose_jacobian()` behavior, my answer is: 😱😱😱

I don't remember why I changed the method's behavior! As you rightly noticed, I implemented it almost ten years ago, and it's hard to remember that decision because I did it all by myself at a time when I didn't have the practice of documenting the rationale behind those decisions. 

Because changing it may have nasty unintended consequences, let's be pragmatic. Please change the `pose_jacobian_derivative()` to match the `pose_jacobian()`.

Best wishes,
Bruno